builders/build/substrate-api/py-substrate-interface.md

```diff
@@ -0,0 +1,311 @@
+
+```python
+# Import
+from substrateinterface import SubstrateInterface
+
+# Construct the API provider
+ws_provider = SubstrateInterface(
+    url="{{ networks.moonbase.wss_url }}",
+)   
+
+# Retrieve the latest block
+block = ws_provider.get_block()
+
+# Retrieve the latest finalized block
+block = ws_provider.get_block_header(finalized_only = True)
+
+# Retrieve a block given its hash
+block_hash = "0xa499d4ebccdabe31218d232460c0f8b91bd08f72aca25f9b25b04b6dfb7a2acb"
+block = ws_provider.get_block(block_hash=block_hash)
+
+for extrinsic in block['extrinsics']:
+
+    if 'address' in extrinsic:
+        signed_by_address = extrinsic['address'].value
+    else:
+        signed_by_address = None
+
+    print('\nPallet: {}\nCall: {}\nSigned by: {}'.format(
+        extrinsic["call"]["call_module"].name,
+        extrinsic["call"]["call_function"].name,
+        signed_by_address
+    ))
+
+```
+
+!!! note
+    The block hash used in the above code sample is the Substrate block hash. The standard methods in Python Substrate Interface assume you are using the Substrate version of primitives, such as block or tx hashes.
+
+
+#### Subscribe to New Block Headers {: #subscribe-to-new-block-headers }
+
+We can also adapt the previous example to use a subscription based model to listen to new block headers.
+
+```python
+# Import
+from substrateinterface import SubstrateInterface
+
+# Construct the API provider
+ws_provider = SubstrateInterface(
+    url="{{ networks.moonbase.wss_url }}",
+) 
+
+def subscription_handler(obj, update_nr, subscription_id):
+
+    print(f"New block #{obj['header']['number']} produced by {obj['author']}")
+
+    if update_nr > 10:
+        return {'message': 'Subscription will cancel when a value is returned', 'updates_processed': update_nr}
+
+
+block_header = ws_provider.subscribe_block_headers(subscription_handler, include_author=True)
+```
+
+### Query for Storage Information {: #query-for-storage-information }
+
+You can use the [`get_metadata_storage_functions`](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.get_metadata_storage_functions){target=_blank} to see a list of available storage functions within Moonbeam network's metadata. 
+
+Chain states that are provided in the metadata through storage functions can be queried through the [`query`](https://polkascan.github.io/py-substrate-interface/#substrateinterface.SubstrateInterface.query){target=_blank} method. 
+
+System modules can be queried to provide basic information such as account nonce and balance. The available functions are read from the metadata dynamically, so it's also possible to query for storage information on Moonbeam custom modules such as `ParachainStaking` for state information that's specific to Moonbeam. 
```

Comment:
A brief description of what are classified as system modules might be helpful or list a couple examples 🤷‍♀️

Reply:
It's mainly just one "System" module. I'll clarify this.